### PR TITLE
HyperRAM test software

### PR DESCRIPTION
**Pull request by alees24**
This PR extends the existing HyperRAM test software and adds new tests specific to the recent HyperRAM development work.

Since the HyperRAM interface is now considerably more involved, retaining data in read buffers (both Instruction and Data ports) as well as a write burst that is under construction, additional tests are introduced to catch edge cases and to test for any coherency issues, both between write and read traffic,as well as between the Instruction and Data ports.

To this end a couple of the tests make extensive use of handwritten assembler to achieve the bus traffic and performance. Compiled code was used initially but this results in well-separated individual read or write transactions and thus does not stress the logic.

_A note on coding style_

Those tests that remain in C++ have also been carefully expressed in places to ensure that the traffic occurs in a short time interval, because otherwise - for example - the hardware will have flushed write traffic out to memory and the test would not be exercising read-write address collisions as intended. In particular note that the pseudo-random number generation is a slow process at the level of CPU cycles/bus transactions, and must thus be avoided between bus transactions that must occur close together.

A brief overview of the modified hardware:

- Data port has a number of 32-byte read buffers, plus one under-construction write burst. Contiguous, ascending write accesses can be coalesced into a burst write of up to 32 bytes. Contiguous descending writes can be coalesced up to a total of 8 bytes (2 words).
- Instruction port has a number of 32-byte read buffers.
- Writes to addresses that lie within the buffered read data will cause the data to be updated (both Data and Instruction ports exhibit this coherency).
- After writing a block of code into the memory, a non-contiguous write is required to ensure that the write data is flushed out, if the code is to be executed immediately and the code is short. (Only 32 bytes may be retained in the construction of a burst write, and write data is flushed out after tens of clock cycles, rather than being retained indefinitely.)

The tests added are as follows:

- performance test; memory-to-memory copy using bursts.
- write tests; specialised implementations of `memset`-like memory initialisation code to exercise the write coalescing logic of the HyperRAM controller interface.
- alignment tests; exercise all possible alignments of source and destination addresses, to check the behaviour of wrapping read bursts and coalesced linear write bursts.
- buffering test, issuing pseudo-random write traffic into a destination buffer with known content.

Sample output from the current HyperRAM implementation (FPGA):
```
Get hyped for hyperram!
Running RND cap test...PASS!
Running RND data test...PASS!
Running RND data & address test...PASS!
Running 0101 stripe test... (7340114 cycles)...PASS!
Running 1001 stripe test... (7340123 cycles)...PASS!
Running 0000_1111 stripe test... (7340119 cycles)...PASS!
Running Execution test... (15666057 cycles)...PASS!
Running performance test with icache enabled...
    copy:   17054 - cmp:  24632 - total:  41686...    copy:   18538 - cmp:  24633 - total:  43171...PASS!
Running performance test with icache disabled...
    copy:   17071 - cmp:  24883 - total:  41954...    copy:   42925 - cmp:  24887 - total:  67812...PASS!
Running alignment tests with cleaning...
without cleaning...
PASS!
```

Sample output from the modified HyperRAM implementation (FPGA) - to be raised in another PR.
```
Get hyped for hyperram!
Running RND cap test...PASS!
Running RND data test...PASS!
Running RND data & address test...PASS!
Running 0101 stripe test... (4947986 cycles)...PASS!
Running 1001 stripe test... (4947987 cycles)...PASS!
Running 0000_1111 stripe test... (4947985 cycles)...PASS!
Running Execution test... (9792557 cycles)...PASS!
Running performance test with icache enabled...
    copy:    3754 - cmp:   6327 - total:  10081...    copy:    3825 - cmp:   6325 - total:  10150...PASS!
Running performance test with icache disabled...
    copy:    3754 - cmp:   6194 - total:   9948...    copy:    3819 - cmp:   6194 - total:  10013...PASS!
Running alignment tests with cleaning...
without cleaning...
PASS!
```

The test `checks/hyperram_test` has been modified to run indefinitely until one or more failures is observed within an iteration, at which point it will halt. It may thus be used to soak test an FPGA build overnight and has run throughout a couple of nights without failure.

### CODE REVIEW

File: sw/cheri/checks/hyperram_test.cc [hunk 1]
```diff
@@ -225,6 +242,376 @@ int execute_test(Capability<volatile uint32_t> &hyperram_area, ds::xoroshiro::P6
...
+        case 1u:
+          burst_hp[wr_offset >> 1] = (uint16_t)d;
+          break;
+        default:
+          burst_wp[wr_offset >> 2] = d;
+          break;
+      }
+    }
+
+    // ----- Avoid the use of randomisation before this point -----
```
Comment (elliotb-lowrisc) [inline sw/cheri/checks/hyperram_test.cc:352]: Is there any use of randomisation after this point?

**Review by elliotb-lowrisc — APPROVED** [review]
Looks good to me. I appreciated the chance to read some assembly functions.

File: sw/cheri/checks/hyperram_test.cc (hunk 1 above)
Comment (alees24) [reply sw/cheri/checks/hyperram_test.cc:352]: Not presently, at least not on that loop iteration. The point was to demarcate - above and below - the region where it must be avoided.

File: sw/cheri/common/hyperram_perf_test.S [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  .p2align 5
+  .type icache_enabled_set,@function
+
+// Set the enabled/disabled state of the icache.
+//
+// a0 = 1 to enable, 0 to disable.
+icache_enabled_set:
+  andi  a0, a0, 1
+	csrw  0x7c0, a0
+	cret
```
[the remark this answers is not in the document]
Comment (alees24) [reply sw/cheri/common/hyperram_perf_test.S:210]: Heh; what's happened there is that I've copied some code from another .S file that uses tabs instead of spaces, and the tab characters are not visible in the editor I used.

File: sw/cheri/common/hyperram_perf_test.S [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  // of the HyperRAM controller and is designed to minimise thrashing on the
+  // LSU read buffer; read 16 bytes of data from one buffer into the register
+  // file and only then start reading from the other and performing comparisons.
+  //
+  // Note: We could replace this with a more conventional implementation now
+  // and compare a larger number of bytes per iteration.
+  clw   a4,   (ca1)
+  clw   a5,  4(ca1)
+  clw   t0,  8(ca1)
+  clw   t1, 12(ca1)  // End of 16 bytes from first buffer.
```
[the remark this answers is not in the document]
Comment (alees24) [reply sw/cheri/common/hyperram_perf_test.S:130]: Fair enough, I'll change that. It just meant the first buffer to be read, rather than how they were named in the function API.
